internal/topology/variables/cluster_variable_defaulting.go

```diff
@@ -73,31 +79,73 @@ func DefaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
 }
 
 // defaultClusterVariable defaults a clusterVariable based on the default value in the clusterClassVariable.
-func defaultClusterVariable(clusterVariable *clusterv1.ClusterVariable, clusterClassVariable *clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+func defaultClusterVariable(clusterVariable *clusterv1.ClusterVariable, clusterClassVariable *clusterv1.ClusterClassVariable, fldPath *field.Path) (*clusterv1.ClusterVariable, field.ErrorList) {
+	// Return if the variable does not exist yet and there is no default value.
+	if clusterVariable == nil && clusterClassVariable.Schema.OpenAPIV3Schema.Default == nil {
+		return nil, nil
+	}
+
 	// Convert schema to Kubernetes APIExtensions schema.
 	apiExtensionsSchema, err := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema)
 	if err != nil {
-		return field.ErrorList{field.Invalid(fldPath, "",
+		return nil, field.ErrorList{field.Invalid(fldPath, "",
 			fmt.Sprintf("invalid schema in ClusterClass for variable %q: error to convert schema %v", clusterVariable.Name, err))}
 	}
 
+	// If it's a new variable, default it.
 	var value interface{}
-	// If the schema has a default value, default it.
-	// NOTE: This is essentially the same which the structural schema defaulting
-	// lib would do, except that it wouldn't default nullable values.
-	// As we want to also default nullable variables we have to do it ourselves.
-	if apiExtensionsSchema.Default != nil {
-		value = runtime.DeepCopyJSONValue(*apiExtensionsSchema.Default)
+	if clusterVariable == nil {
+		// TODO(sbueringer): this if branch can be deleted when we drop top-level nullable.
+		if apiExtensionsSchema.Default != nil {
+			// NOTE: This is essentially the same what the structural schema defaulting
+			// lib would do, except that it wouldn't default nullable values.
+			// As we want to also default nullable variables we have to do it ourselves.
+			value = runtime.DeepCopyJSONValue(*apiExtensionsSchema.Default)
+		}
```

Comment:
```suggestion
```
+ corresponding unit tests will be done once https://github.com/kubernetes-sigs/cluster-api/pull/5780 has been merged